tapecntl/tapecntl.1
.\"
.\" Sccsid @(#)tapecntl.1	1.6 (gritter) 12/9/04
.TH TAPECNTL 1 "12/9/04" "Heirloom Toolchest" "User Commands"
.SH NAME
tapecntl \- tape control for tape devices
.SH SYNOPSIS
\fBtapecntl\fR
[\fB\-etrwaluv\fR]
[\fB\-c\ \fIarg\fR]
[\fB\-p\ \fIarg\fR]
[\fB\-d\ \fIdensity_in_decimal\fR]
[\fB\-f\ \fIarg\fR]
[\fIdevice\fR]
.SH DESCRIPTION
The
.I tapecntl
command performs certain control operations on tape devices.
If neither a
.I device
argument is present
nor the
.I TAPE
environment variable is set,
the device /dev/rmt/c0s0n is used.
.PP
The following options are accepted:
.TP
.B \-e
Erase the entire tape.
.TP
.B \-t
Retension the tape,
i.\|e. rewind the tape,
wind the tape to end of reel
and rewind it.
.TP
.B \-r
Reset the tape device.
.TP
.B \-w
Rewind the tape.
.TP
\fB\-p\ \fIposition\fR
Position the tape
at the data after the named file mark.
The
.I position
argument is interpreted relative to the current tape position.
.PP
The following options are extensions:
.TP
.B \-a
Position the tape at end of data.
.TP
.B \-l
Load the tape.
.TP
.B \-u
Unload the tape.
.TP
\fB\-d\ \fIdensity_in_decimal\fR
Set the SCSI tape density.
Some density codes are:
0 = default (autodetect);
4 = QIC-11;
5 = QIC-24;
15 = QIC-120;
16 = QIC-150;
17 = QIC-320;
20 = EXB-8200;
21 = EXB-8500;
21 = QIC-1000;
34 = QIC-2GB;
140 = EXB-8500c;
144 = EXB-8200c.
.TP
\fB\-c\ \fIarg\fR
Enable compression (if \fIarg\fR is different from \fB0\fR)
or disable it (if \fIarg\fR equals \fB0\fR).
.TP
\fB\-f\ \fIarg\fR
Set the tape block size to
.I arg
bytes.
.TP
.B \-v
Set variable tape block size.
.SH "ENVIRONMENT VARIABLES"
.TP
TAPE
Used as tape device
if no
.I device
argument is given on the command line.


CHANGES
Release 070715
* cpio: A "-P" option has been introduced that allows to specify attributes
  like the file owner when creating an archive. With it, unprivileged users
  can set arbitrary attributes for files within an archive.
* diff: Made it compile on CentOS 5.
* Static linking against glibc is now possible since "optopt" is no longer
  a common symbol, and there is no conflict anymore with "getopt.o".
* cpio, tar: A compilation problem on some Linux systems has been fixed.
* ul: A compilation problem for NetBSD and other systems that use termcap
  instead of curses has been fixed (patch by Aleksey Cheusov).
* bfs: An OpenSolaris-derived manual page is now included.

Release 070227
* The directory for development utilities, by default "/usr/ccs/bin", is
  now configurable in "build/mk.config" . It is expected that the utilities
  from the "Heirloom Development Tools" package are installed in it.
* getconf: "/usr/ccs/bin" is now included in the default path as printed
  with the "PATH" argument.
* what: Has been replaced by an OpenSolaris-derived implementation.
* tsort: Is now installed in "/usr/ccs/bin".
* OpenSolaris-derived manual pages are now included for bdiff, newform,
  and fspec(5).
* psrinfo: Fixed the computation of the "on-line since" time for Linux 2.6.
* cpio: Added a "-D" option to make it usable in combination with SVR4
  packaging tools. This option is not intended for general use, and is
  thus not documented.
* The package build target "make pkg" for Solaris and Open UNIX is no
  longer supported.

Release 070115
* deroff: Now recognizes the ".do" and ".xflag" requests as well as an "-xN"
  command line option to handle Heirloom nroff/troff extensions.
* ed: Substituting only a specific occurence of the pattern, as in the
  command "s/x/y/3", now works correctly if the substitution "y" is not
  equal in length to the pattern match "x" (Derek Thomas).
* ps: Use the Linux 2.6 /proc/status "SleepAVG" field to compute the value
  for the "C" column (cpu usage for scheduling).
* priocntl: The "FF", "B", and "ISO" class names are now accepted with
  the "-e" option too.
* file: UTF-8 characters in troff input files are now recognized.
* cp: The "-s" option now displays correct sizes and related statistics
  for special files with st_size == 0.
* grep: With the "-z" option, decompressor children processes are now
  properly waited for, avoiding failures when processing many files.
* su: Works with PAM on Mac OS X now (Stefan Tramm).
* ps: Handles dead processes as on other platforms now (Stefan Tramm).
* pgrep: Ignores dead processes now (Stefan Tramm).
* spell: A value of the wrong type that was passed to scanf() could lead
  to build errors on 64-bit platforms (Jeffrey H. Johnson).
* ps, pgrep, whodo: Works on FreeBSD 6 now.
* df, find, ps: Changes have been made to build on NetBSD 3 (Aleksey Cheusov).
* Changes have been made to build on OpenBSD 3.8 (Aleksey Cheusov).
* Changes have been made to build with GNU make 3.81.

Release 060122
* The tools have been ported to Mac OS X 10.4.3 (patches by Stefan Tramm).
* The tools have been ported to DragonFlyBSD 1.3.7-DEVELOPMENT.

Release 051118
* rm: The output of the question before descending to a directory with
  the "-i" option has been fixed.
* /usr/ucb/ps: The "-a" option has been fixed; it incorrectly printed a
  list of all processes since release 050411.
* getconf: /usr/ucb is now put at the end of the Heirloom directory list
  for the PATH variable.
* The "-man" macro no longer produces cut marks when used with troff.
* tr: When evaluating '[:class:]' expressions, the last position of the
  coded character set was not included; this has been fixed.
* ps: The "-R" option (resource usage format), which had been present
  for some time, has been documented.
* RPM package builds now require the new "Heirloom Development Tools"
  package from <http://heirloom.sourceforge.net/devtools.html> to be
  installed.

Release 050815
* The "getconf" utility has been added (derived from OpenSolaris code).
* The "mesg" and "pathchk" utilities have been added.
* tr: A new implementation that is based on 6th and 7th edition code is
  now used. It stores the translation tables in arrays of variable size.
  This results in better performance, under some circumstances at the
  expense of higher memory consumption.
* printf: Handle unsigned values larger than INT_MAX correctly with "%u",
  "%o", "%x", and "%X".
* printf: Handle SVID3 "position" fields in format specifications, such
  as "%7$s", that specify which argument is used next.
* pg: If a line does not fit on the screen, do not print a newline to
  force wraparound unless it is really necessary. This allows to select
  words across the wrap for copy-and-paste in an xterm.
* pg: Discard the ESC[48m sequence on a terminal that supports colors.
* ed: Do not print "expecting `w'" with the "q" command after a zero-byte
  file has been written.
* man: The tmac.an macros no longer use "@" as an escape character; its
  previous use lead to weird output with Unix nroff when an "@" appeared
  in titles or hanging labels.
* man: The "TCAT" post processor is now invoked separately for each manual
  entry. Previously, when the -t option was given with multiple entries and
  TCAT=dpost, only the first entry was actually printed.
* Some issues with building on AIX have been fixed.

Release 050713
* The 'echo' and 'test' commands have been added as standalone versions of
  those of the Heirloom Bourne shell.
* The 'false', 'kill', 'printf', 'pwd', and 'true' commands have been added.

tapecntl/tapecntl.c
/*
 * tapecntl - tape control for tape devices
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, December 2003.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 *
 * This software is provided 'as-is', without any express or implied
 * warranty. In no event will the authors be held liable for any damages
 * arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute
 * it freely, subject to the following restrictions:
 *
 * 1. The origin of this software must not be misrepresented; you must not
 *    claim that you wrote the original software. If you use this software
 *    in a product, an acknowledgment in the product documentation would be
 *    appreciated but is not required.
 *
 * 2. Altered source versions must be plainly marked as such, and must not be
 *    misrepresented as being the original software.
 *
 * 3. This notice may not be removed or altered from any source distribution.
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
static const char sccsid[] USED = "@(#)tapecntl.sl	1.38 (gritter) 1/22/06";

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <libgen.h>
#include <errno.h>
#if defined (__linux__) || defined (__sun) || defined (__FreeBSD__) \
	|| defined (__hpux) || defined (_AIX) || defined (__NetBSD__) \
	|| defined (__OpenBSD__) || defined (__DragonFly__) \
	|| defined (__APPLE__)
#include <sys/ioctl.h>
#include <sys/mtio.h>
#else	/* SVR4.2MP */
#include <sys/ioctl.h>
#include <sys/scsi.h>
#include <sys/st01.h>
#endif	/* SVR4.2MP */

#define	eq(a, b)	(strcmp(a, b) == 0)

static const char	*tape;
static const char	deftape[] = "/dev/rmt/c0s0n";
static const char	*progname;
static int		status;
static int		omode = O_RDONLY;

static int		mt;

static enum oper {
	OP_NONE		= 0,
	OP_OFFLINE	= 0000001,
	OP_UNLOAD	= 0000002,
	OP_LOAD		= 0000004,
	OP_RESET	= 0000010,
	OP_RETENSION	= 0000020,
	OP_REWIND	= 0000040,
	OP_COMPRESSION	= 0000100,
	OP_DENSITY	= 0000200,
	OP_FIXED	= 0000400,
	OP_ERASE	= 0001000,
	OP_POSITION	= 0002000,
	OP_BACKWARD	= 0004000,
	OP_FWBLOCKS	= 0010000,
	OP_BWBLOCKS	= 0020000,
	OP_ENDDATA	= 0040000,
	OP_WRITEMARK	= 0100000
} operations;

static int		verbose;
static const char	*stats_info;
static int		stats_flags_ok;
static long		stats_flags;
static long		stats_density = -1;
static long		stats_recsiz = -1;

static int		count;

static void
m_usage(void)
{
	fprintf(stderr, "Usage: %s [-v] [-f tapename] command [count]\n",
			progname);
	fprintf(stderr, "Default device is %s\n", deftape);
	fprintf(stderr, "Common Tape Commands:\n"
			"  erase         - erase entire tape\n"
			"  rewind        - rewind the tape\n"
			"  retension     - wind tape to end of reel "
				"and rewind it\n"
			"Other Tape Commands:\n"
			"  eof, wfm      - write file mark\n"
			"  fsf, sfm      - seek next file mark\n"
			"  bsf           - seek previous file mark\n"
			"  enddata       - seek end of data\n");
	exit(1);
}

static void
m_options(int argc, char **argv)
{
	const char	optstring[] = ":vf:";
	int	i;

	while ((i = getopt(argc, argv, optstring)) != EOF) {
		switch (i) {
		case 'v':
			verbose = 1;
			omode |= O_NONBLOCK;
			break;
		case 'f':
			tape = optarg;
			break;
		default:
			m_usage();
		}
	}
	if (tape == NULL) {
		if ((tape = getenv("TAPE")) != NULL)
			tape = strdup(tape);
		else
			tape = deftape;
	}
	if (optind >= argc) {
		if (verbose)
			return;
		m_usage();
	}
	if (eq(argv[optind], "erase")) {
		operations = OP_ERASE;
		omode = O_RDWR | omode&O_NONBLOCK;
	} else if (eq(argv[optind], "rewind"))
		operations = OP_REWIND;
	else if (eq(argv[optind], "retension") || eq(argv[optind], "ret"))
		operations = OP_RETENSION;
	else if (eq(argv[optind], "wfm") || eq(argv[optind], "eof") ||
			eq(argv[optind], "weof")) {
		operations = OP_WRITEMARK;
		omode = O_RDWR | omode&O_NONBLOCK;
		if (optind+1 < argc)
			count = atoi(argv[optind+1]);
		else
			count = 1;
	} else if (eq(argv[optind], "sfm") || eq(argv[optind], "fsf")) {
		operations = OP_POSITION;
		if (optind+1 < argc)
			count = atoi(argv[optind+1]);
		else
			count = 1;
	} else if (eq(argv[optind], "bsf")) {
		operations = OP_BACKWARD;
		if (optind+1 < argc)
			count = atoi(argv[optind+1]);
		else
			count = 1;
	} else if (eq(argv[optind], "fsr")) {
		operations = OP_FWBLOCKS;
		if (optind+1 < argc)
			count = atoi(argv[optind+1]);
		else
			count = 1;
	} else if (eq(argv[optind], "bsr")) {
		operations = OP_BWBLOCKS;
		if (optind+1 < argc)
			count = atoi(argv[optind+1]);
		else
			count = 1;
	} else if (eq(argv[optind], "compression")) {
		operations = OP_COMPRESSION;
		if (optind+1 < argc)
			count = atoi(argv[optind+1]);
		else
			count = 1;
	} else if (eq(argv[optind], "enddata") || eq(argv[optind], "append") ||
			eq(argv[optind], "feom"))
		operations = OP_ENDDATA;
	else if (eq(argv[optind], "offline") || eq(argv[optind], "rewoffl"))
		operations = OP_OFFLINE;
	else if (eq(argv[optind], "status")) {
		verbose = 1;
		omode |= O_NONBLOCK;
	} else
		m_usage();
}

static void
tape_usage(void)
{
	fprintf(stderr, "\
tape v%s usage:   %s [-<tape>] [-a arg] <command> [device]\n\
  SCSI tape commands:\n\
        status     - get the status of the tape\n\
        reten      - retension the tape\n\
        erase      - erase and retension tape\n\
        reset      - reset controller and driver\n\
        rewind     - rewind tape controller\n\
        rfm        - skip to next file\n\
        eod        - position to EOD\n\
        wfm        - write filemark\n\
        load       - load tape\n\
        unload     - unload tape\n\
        setblk     - set block size (in bytes) for device\n\
        setcomp    - set compression (0 disabled, 1 enabled)\n\
        setdensity - set density code (in hexadecimal)\n",
        "1.38",
        progname);
        exit(1);
}

static const char *
tape_default_device(void)
{
	FILE	*fp;
	int	c, t = 0;
	const char	sts[] = " device =";
	char	*device;

	if ((device = getenv("TAPE")) != NULL)
		device = strdup(device);
	else
		device = (char *)deftape;
	if ((fp = fopen(TAPEDFL, "r")) == NULL)
		return device;
	while ((c = getc(fp)) != EOF) {
		switch (c) {
		case '\n':
			t = 0;
			break;
		case ' ':
			if (c != ' ')
				t = -1;
			break;
		default:
			if (t >= 0) {
				if (sts[t] == ' ')
					t++;
				if (sts[t] == c)
					t++;
				else
					t = -1;
			}
		}
		printf("c = %c t = %d\n", c, t);
		if (t >= 0 && sts[t] == '\0') {
			while ((c = getc(fp)) == ' ');
			t = 0;
			device = malloc(1);
			while (c != EOF && c != '\n') {
				device[t++] = c;
				device = realloc(device, t + 1);
				c = getc(fp);
			}
			device[t] = '\0';
		}
	}
	fclose(fp);
	return device;
}

static void
tape_options(int argc, char **argv)
{
	const char	optstring[] = ":csf8ia:";
	const char	*aflag = 0;
	int	i;

	while ((i = getopt(argc, argv, optstring)) != EOF) {
		switch (i) {
		case 'a':
			aflag = optarg;
			break;
		case '?':
		case ':':
			fprintf(stderr, " %s: unknown option '-%c'\n",
					progname, i);
			tape_usage();
		}
	}
	if (optind == argc)
		tape_usage();
	if (eq(argv[optind], "status")) {
		verbose = 1;
		omode |= O_NONBLOCK;
	} else if (eq(argv[optind], "reten"))
		operations |= OP_RETENSION;
	else if (eq(argv[optind], "erase")) {
		operations |= OP_ERASE;
		omode = O_RDWR | omode&O_NONBLOCK;
	} else if (eq(argv[optind], "reset")) {
		operations |= OP_RESET;
		omode |= O_NONBLOCK;
	} else if (eq(argv[optind], "rewind"))
		operations |= OP_REWIND;
	else if (eq(argv[optind], "rfm")) {
		operations |= OP_POSITION;
		count = aflag ? atoi(aflag) : 1;
	} else if (eq(argv[optind], "eod")) {
		operations |= OP_ENDDATA;
	} else if (eq(argv[optind], "wfm")) {
		operations |= OP_WRITEMARK;
		omode = O_RDWR | omode&O_NONBLOCK;
		count = aflag ? atoi(aflag) : 1;
	} else if (eq(argv[optind], "load")) {
		operations |= OP_LOAD;
		omode |= O_NONBLOCK;
	} else if (eq(argv[optind], "unload"))
		operations |= OP_UNLOAD;
	else if (eq(argv[optind], "setblk")) {
		operations |= OP_FIXED;
		count = aflag ? atoi(aflag) : 0;
	} else if (eq(argv[optind], "setcomp")) {
		operations |= OP_COMPRESSION;
		count = aflag ? atoi(aflag) : 1;
	} else if (eq(argv[optind], "setdensity")) {
		operations |= OP_DENSITY;
		if (aflag && aflag[0] == '0' &&
				(aflag[1] == 'x' || aflag[1] == 'X'))
			aflag += 2;
		count = aflag ? strtol(aflag, NULL, 16) : 0;
	} else {
		fprintf(stderr, " %s: invalid command '%s'\n",
				progname, argv[optind]);
		tape_usage();
	}
	if (optind < argc-1)
		tape = argv[optind+1];
	else
		tape = tape_default_device();
}

static void
t_options(int argc, char **argv)
{
	const char	optstring[] = "etrwp:alud:f:c:v";
	int	i;

	while ((i = getopt(argc, argv, optstring)) != EOF) {
		switch (i) {
		case 'e':
			operations |= OP_ERASE;
			omode = O_RDWR | omode&O_NONBLOCK;
			break;
		case 't':
			operations |= OP_RETENSION;
			break;
		case 'r':
			operations |= OP_RESET;
			omode |= O_NONBLOCK;
			break;
		case 'w':
			operations |= OP_REWIND;
			break;
		case 'p':
			operations |= OP_POSITION;
			count = atoi(optarg);
			break;
		case 'a':
			operations |= OP_ENDDATA;
			break;
		case 'l':
			operations |= OP_LOAD;
			omode |= O_NONBLOCK;
			break;
		case 'u':
			operations |= OP_UNLOAD;
			break;
		case 'd':
			operations |= OP_DENSITY;
			count = atoi(optarg);
			break;
		case 'f':
			operations |= OP_FIXED;
			count = atoi(optarg);
			break;
		case 'v':
			operations |= OP_FIXED;
			count = 0;
			break;
		case 'c':
			operations |= OP_COMPRESSION;
			count = atoi(optarg);
			break;
		default:
		usage:	fprintf(stderr, "Usage: %s [ -etrw ] [ -p arg ]\n",
					progname);
			exit(1);
		}
	}
	if (operations == OP_NONE)
		goto usage;
	if (optind < argc)
		tape = argv[optind];
	else if ((tape = getenv("TAPE")) != NULL)
		tape = strdup(tape);
	else
		tape = deftape;
}

#if defined (__linux__)
static int
process(int fd, enum oper op)
{
	struct mtop	m;

	switch (op) {
	default:
		errno = ENOSYS;
		return -1;
	case OP_RESET:
		m.mt_op = MTRESET;
		break;
	case OP_RETENSION:
		m.mt_op = MTRETEN;
		break;
	case OP_REWIND:
		m.mt_op = MTREW;
		break;
	case OP_POSITION:
		m.mt_op = MTFSF;
		m.mt_count = count;
		break;
	case OP_BACKWARD:
		m.mt_op = MTBSF;
		m.mt_count = count;
		break;
	case OP_FWBLOCKS:
		m.mt_op = MTFSR;
		m.mt_count = count;
		break;
	case OP_BWBLOCKS:
		m.mt_op = MTBSR;
		m.mt_count = count;
		break;
	case OP_ERASE:
		m.mt_op = MTERASE;
		break;
	case OP_WRITEMARK:
		m.mt_op = MTWEOF;
		m.mt_count = count;
		break;
	case OP_ENDDATA:
		m.mt_op = MTEOM;
		break;
	case OP_OFFLINE:
		m.mt_op = MTOFFL;
		break;
	case OP_LOAD:
		m.mt_op = MTLOAD;
		break;
	case OP_UNLOAD:
		m.mt_op = MTUNLOAD;
		break;
	case OP_DENSITY:
		m.mt_op = MTSETDENSITY;
		m.mt_count = count;
		break;
	case OP_FIXED:
		m.mt_op = MTSETBLK;
		m.mt_count = count;
		break;
	case OP_COMPRESSION:
		m.mt_op = MTCOMPRESSION;
		m.mt_count = count;
		break;
	}
	return ioctl(fd, MTIOCTOP, &m);
}

static int
stats(int fd)
{
	struct mtget	mg;
	static struct {
		int	type;
		const char	*name;
	} ti[] = MT_TAPE_INFO;
	int	i;

	if (ioctl(fd, MTIOCGET, &mg) < 0)
		return -1;
	for (i = 0; ti[i].name; i++)
		if (ti[i].type == mg.mt_type) {
			stats_info = ti[i].name;
			break;
		}
	stats_flags_ok = 1;
	stats_flags = mg.mt_gstat;
	stats_density = (mg.mt_dsreg&MT_ST_DENSITY_MASK)>>MT_ST_DENSITY_SHIFT;
	stats_recsiz = (mg.mt_dsreg&MT_ST_BLKSIZE_MASK) >> MT_ST_BLKSIZE_SHIFT;
	return 0;
}
#elif defined (__sun)
static int
process(int fd, enum oper op)
{
	struct mtop	m;

	switch (op) {
	default:
		errno = ENOSYS;
		return -1;
	case OP_RETENSION:
		m.mt_op = MTRETEN;
		break;
	case OP_REWIND:
		m.mt_op = MTREW;
		break;
	case OP_POSITION:
		m.mt_op = MTFSF;
		m.mt_count = count;
		break;
	case OP_BACKWARD:
		m.mt_op = MTBSF;
		m.mt_count = count;
		break;
	case OP_FWBLOCKS:
		m.mt_op = MTFSR;
		m.mt_count = count;
		break;
	case OP_BWBLOCKS:
		m.mt_op = MTBSR;
		m.mt_count = count;
		break;
	case OP_WRITEMARK:
		m.mt_op = MTWEOF;
		m.mt_count = count;
		break;
	case OP_ENDDATA:
		m.mt_op = MTEOM;
		break;
	case OP_OFFLINE:
		m.mt_op = MTOFFL;
		break;
	case OP_LOAD:
		m.mt_op = MTLOAD;
		break;
	case OP_FIXED:
		m.mt_op = MTSRSZ;
		m.mt_count = count;
		break;
	case OP_ERASE:
		m.mt_op = MTERASE;
		break;
	}
	return ioctl(fd, MTIOCTOP, &m);
}

static int
stats(int fd)
{
	struct mtdrivetype_request	mr;
	static struct mtdrivetype	md;

	mr.size = sizeof md;
	mr.mtdtp = &md;
	if (ioctl(fd, MTIOCGETDRIVETYPE, &mr) < 0)
		return -1;
	stats_info = md.name;
	stats_recsiz = md.bsize;
	stats_flags_ok = 1;
	stats_flags = md.options;
	stats_density = md.default_density;
}
#elif defined (__FreeBSD__) || defined (__NetBSD__) || defined (__OpenBSD__) \
	|| defined (__DragonFly__) || defined (__APPLE__)
static int
process(int fd, enum oper op)
{
	struct mtop	m;

	switch (op) {
	default:
		errno = ENOSYS;
		return -1;
#ifdef	MTRETENS
	case OP_RETENSION:
		m.mt_op = MTRETENS;
		break;
#endif
	case OP_REWIND:
		m.mt_op = MTREW;
		break;
	case OP_POSITION:
		m.mt_op = MTFSF;
		m.mt_count = count;
		break;
	case OP_BACKWARD:
		m.mt_op = MTBSF;
		m.mt_count = count;
		break;
	case OP_FWBLOCKS:
		m.mt_op = MTFSR;
		m.mt_count = count;
		break;
	case OP_BWBLOCKS:
		m.mt_op = MTBSR;
		m.mt_count = count;
		break;
	case OP_ERASE:
		m.mt_op = MTERASE;
		break;
	case OP_WRITEMARK:
		m.mt_op = MTWEOF;
		m.mt_count = count;
		break;
#ifdef	MTEOD
	case OP_ENDDATA:
		m.mt_op = MTEOD;
		break;
#endif
	case OP_OFFLINE:
		m.mt_op = MTOFFL;
		break;
	case OP_DENSITY:
		m.mt_op = MTSETDNSTY;
		m.mt_count = count;
		break;
		break;
	case OP_FIXED:
		m.mt_op = MTSETBSIZ;
		m.mt_count = count;
		break;
#ifdef	MTCOMP
	case OP_COMPRESSION:
		m.mt_op = MTCOMP;
		m.mt_count = count;
		break;
#endif
	}
	return ioctl(fd, MTIOCTOP, &m);
}

static int
stats(int fd)
{
	struct mtget	mg;

	if (ioctl(fd, MTIOCGET, &mg) < 0)
		return -1;
	stats_flags_ok = 1;
	stats_flags = mg.mt_dsreg;
	stats_density = mg.mt_density;
	stats_recsiz = mg.mt_blksiz;
	return 0;
}
#elif defined (__hpux) || defined (_AIX)
static int
process(int fd, enum oper op)
{
	struct mtop	m;

	switch (op) {
	default:
		errno = ENOSYS;
		return -1;
	case OP_REWIND:
		m.mt_op = MTREW;
		break;
	case OP_POSITION:
		m.mt_op = MTFSF;
		m.mt_count = count;
		break;
	case OP_BACKWARD:
		m.mt_op = MTBSF;
		m.mt_count = count;
		break;
	case OP_FWBLOCKS:
		m.mt_op = MTFSR;
		m.mt_count = count;
		break;
	case OP_BWBLOCKS:
		m.mt_op = MTBSR;
		m.mt_count = count;
		break;
	case OP_ERASE:
#ifdef	MTERASE
		m.mt_op = MTERASE;
#endif
#ifdef	STERASE
		m.mt_op = STERASE;
#endif
		break;
	case OP_WRITEMARK:
		m.mt_op = MTWEOF;
		m.mt_count = count;
		break;
#ifdef	MTEOD
	case OP_ENDDATA:
		m.mt_op = MTEOD;
		break;
#endif
	case OP_OFFLINE:
		m.mt_op = MTOFFL;
		break;
	}
	return ioctl(fd, MTIOCTOP, &m);
}

static int
stats(int fd)
{
	struct mtget	mg;

	if (ioctl(fd, MTIOCGET, &mg) < 0)
		return -1;
	stats_flags_ok = 1;
	return 0;
}
#else	/* SVR4.2MP */
static int
process(int fd, enum oper op)
{
	switch (op) {
	default:
		errno = ENOSYS;
		return -1;
	case OP_RESET:
		return ioctl(fd, T_RST);
	case OP_RETENSION:
		return ioctl(fd, T_RETENSION);
	case OP_REWIND:
		return ioctl(fd, T_RWD);
	case OP_POSITION:
		return ioctl(fd, T_SFF, count);
	case OP_BACKWARD:
		return ioctl(fd, T_SFB, count);
	case OP_FWBLOCKS:
		return ioctl(fd, T_SBF, count);
	case OP_BWBLOCKS:
		return ioctl(fd, T_SBB, count);
	case OP_WRITEMARK:
		return ioctl(fd, T_WRFILEM, 1);
	case OP_ENDDATA:
		return ioctl(fd, T_EOD);
	case OP_LOAD:
		return ioctl(fd, T_LOAD);
	case OP_UNLOAD:
		return ioctl(fd, T_UNLOAD);
	case OP_DENSITY:
		return ioctl(fd, T_STD, count);
	case OP_FIXED: {
		struct blklen	bl;
		bl.min_blen = bl.max_blen = count;
		return ioctl(fd, T_WRBLKLEN, &bl);
	}
	case OP_ERASE:
		return ioctl(fd, T_ERASE);
	case OP_COMPRESSION:
		return ioctl(fd, T_SETCOMP, count);
	}
}

static int
stats(int fd)
{
	return -1;
}
#endif	/* SVR4.2MP */

int
main(int argc, char **argv)
{
	int	fd;
	int	i;

	progname = basename(argv[0]);
	if (progname[0] == 'm' && progname[1] == 't') {
		mt = 1;
		m_options(argc, argv);
	} else if (strstr(progname, "cntl") == NULL) {
		tape_options(argc, argv);
	} else
		t_options(argc, argv);
	if (mt) {
		struct stat	st;

		if (stat(tape, &st) < 0) {
			fprintf(stderr, "%s: could not find %s\n"
					"%s: %s\n",
					progname, tape,
					progname, strerror(errno));
			exit(1);
		} else if ((st.st_mode&S_IFMT) != S_IFCHR) {
			fprintf(stderr, "%s: %s is not a character device.\n",
					progname, tape);
			exit(1);
		}
	}
	if ((fd = open(tape, omode)) < 0) {
		if (mt) {
			status |= 1;
			fprintf(stderr, "%s: could not open %s\n"
					"%s: %s\n",
					progname, tape,
					progname, strerror(errno));
		} else if (operations & OP_ERASE && errno == EROFS) {
			status |= 3;
			fprintf(stderr, "Write protected cartridge.\n");
		} else {
			status |= errno == EBUSY ? 4 : 1;
			fprintf(stderr, "Device open failure.\n");
			fprintf(stderr, "Please check that cartridge or device "
					"cables are inserted properly.\n");
		}
		exit(status);
	}
	if (verbose && stats(fd) == 0) {
		if (stats_info != NULL)
			printf("Found %s:\n", stats_info);
		if (stats_flags_ok)
			printf("flags   = 0x%lx\n", stats_flags);
		if (stats_density >= 0)
			printf(mt ? "density = %ld\n" : "density = 0x%lx\n",
					stats_density);
		if (stats_recsiz >= 0)
			printf("recsiz  = %ld\n", stats_recsiz);
	}
	if (operations == OP_NONE)
		exit(0);
	for (i = 1; i <= 0177777; i <<= 1) {
		if (operations & i && process(fd, i) < 0) {
			const char	*msg = "";

			switch (i) {
			case OP_RESET:
				msg = "Reset function failure.";
				break;
			case OP_RETENSION:
				msg = "Retension function failure.\n"
					"Please check equipment carefully "
					"before retry.";
				break;
			case OP_REWIND:
				msg = "Rewind function failure.";
				break;
			case OP_ENDDATA:
			case OP_POSITION:
			case OP_BACKWARD:
			case OP_FWBLOCKS:
			case OP_BWBLOCKS:
				msg = "Positioning function failure";
				break;
			case OP_ERASE:
				if (errno == EACCES)
					msg = "Erase function failure.\n"
						"Write protected cartridge.";
				else
					msg = "Erase function failure.\n"
						"Please check that cartridge "
						"or device cables are "
						"inserted properly";
			case OP_OFFLINE:
				msg = "Offline function failure.";
				break;
			case OP_UNLOAD:
				msg = "Unload function failure.";
				break;
			case OP_LOAD:
				msg = "Load function failure.";
				break;
			case OP_DENSITY:
				msg = "Set tape density function failure.";
				break;
			case OP_FIXED:
				msg = "Set block length function failure.";
				break;
			case OP_WRITEMARK:
				msg = "Write mark function failure.";
			default:
				msg = strerror(errno);
			}
			fprintf(stderr, "%s\n", msg);
			status |= errno == EACCES ? 3 : 2;
		}
	}
	return status;
}
